src/geometry/detail/VGNavCollection.cc

```diff
@@ -0,0 +1,97 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file VGNavCollection.cc
+//---------------------------------------------------------------------------//
+#include "VGNavCollection.hh"
+
+#include <VecGeom/navigation/NavStatePool.h>
+#include "comm/Device.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+// HOST VALUE
+//---------------------------------------------------------------------------//
+void VGNavCollection<Ownership::value, MemSpace::host>::resize(int max_depth,
+                                                               size_type size)
+{
+    CELER_EXPECT(max_depth > 0);
+    CELER_EXPECT(size == 1);
+
+    nav_state.reset(NavState::MakeInstance(max_depth));
+}
+
+//---------------------------------------------------------------------------//
+// HOST REFERENCE
+//---------------------------------------------------------------------------//
+/*!
+ * Get a reference to host value data.
+ */
+void VGNavCollection<Ownership::reference, MemSpace::host>::operator=(
+    VGNavCollection<Ownership::value, MemSpace::host>& other)
+{
+    ptr = other.nav_state.get();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the navigation state at the given thread, which must be zero.
+ */
+auto VGNavCollection<Ownership::reference, MemSpace::host>::at(int,
+                                                               ThreadId id) const
+    -> NavState&
+{
+    CELER_EXPECT(*this);
+    CELER_EXPECT(id.get() == 0);
+    return *ptr;
+}
+
+//---------------------------------------------------------------------------//
+// DEVICE VALUE
+//---------------------------------------------------------------------------//
+/*!
+ * Deleter frees CUDA data.
+ */
+void NavStatePoolDeleter::operator()(arg_type ptr) const
+{
+    delete ptr;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Allocate the pool and save the GPU pointer.
+ */
+void VGNavCollection<Ownership::value, MemSpace::device>::resize(int max_depth,
+                                                                 size_type sz)
+{
+    CELER_EXPECT(max_depth > 0);
+    CELER_EXPECT(sz > 0);
+    CELER_EXPECT(celeritas::device());
+
+    pool.reset(new vecgeom::cxx::NavStatePool(sz, max_depth));
```

Comment:
This is likely related to my previous question.  Why does the host contain a `NavState` and the device a `NavStatePool`?

Reply:
I think I answered your question above -- navstatepool isn't very CPU-friendly.